docs/spec/distributing.rst

```diff
@@ -266,3 +622,23 @@ of that Python version. This can be queried e.g.
 ``pythonX.Y -c 'import site; print(site.getsitepackages())'``. It is also recommended
 that the type checker allow for the user to point to a particular Python
 binary, in case it is not in the path.
+
+.. _import-conventions:
+
+Import Conventions
+------------------
+
+By convention, certain import forms indicate to type checkers that an imported
+symbol is re-exported and should be considered part of the importing module's
+public interface. All other imported symbols are considered private by default.
+
+The following import forms re-export symbols:
+
+* ``import X as X`` (a redundant module alias): re-exports ``X``.
+* ``from Y import X as X`` (a redundant symbol alias): re-exports ``X``.
+* ``from Y import *``: re-exports all symbols in ``Y`` that do not begin with
+  an underscore.
+* ``from . import bar`` in an ``__init__`` module: re-exports ``bar`` if it does
```

Comment:
I don't think these last two bullets should be included. They kind of make sense for stubs, but they don't make sense for "py.typed" libraries. I think the rules for stubs and "py.typed" libraries should be the same. Pyright has never implemented these two rules for either stubs or "py.typed" libraries, and they are not included in [this guide](https://github.com/python/typing/blob/main/docs/guides/libraries.rst#library-interface-public-and-private-symbols).

Reply:
That guide says, "If [imported symbols] use the [...] `from . import A` forms, symbol `A` is not private unless the name begins with an underscore. If a file `__init__.py` uses form `from .A import X`, symbol `A` is treated likewise," which I interpreted to mean what I wrote in the last two bullets. Did I misunderstand something?

Reply:
I do think these last two bullets make sense for `py.typed` libraries. It's common for libraries to define classes and functions in sub-modules, and then import them in `__init__.py` for the sole purpose of exposing a flat namespace to users of the library.

Reply:
I've seen cases where a "py.typed" library includes a `__init__.py` that imports submodules with the intent of referencing specific symbols in those submodules but without the intent of re-exporting that submodule. If we always assume that library authors want these submodules re-exported from the top-level module, we're taking away the ability for them to keep them private. If they want to make them public, they can explicitly do so by using a redundant form (`from . import bar as bar`). Pyright has enforced this for the last three years, and library authors have adapted accordingly. I'm uncomfortable changing the rules now.

Reply:
Ah, that makes a lot of sense. I've removed the last two bullets.